csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll

```diff
@@ -755,588 +755,499 @@ module ControlFlow {
+          not c0 instanceof BooleanCompletion and
+          c = c0
+          or
+          rec = TLastRecBreakCompletion() and
+          c0 instanceof BreakCompletion and
+          c instanceof BreakNormalCompletion
+          or
+          rec = TLastRecSwitchAbnormalCompletion() and
+          not c instanceof BreakCompletion and
+          not c instanceof NormalCompletion and
+          not c instanceof GotoDefaultCompletion and
+          not c instanceof GotoCaseCompletion and
+          c = c0
+          or
+          rec = TLastRecInvalidOperationException() and
+          (c0.(MatchingCompletion).isNonMatch() or c0 instanceof FalseCompletion) and
+          c
+              .(ThrowCompletion)
+              .getExceptionClass()
+              .hasQualifiedName("System.InvalidOperationException")
+          or
+          rec = TLastRecNonContinueCompletion() and
+          not c0 instanceof BreakCompletion and
+          not c0.continuesLoop() and
+          c = c0
+          or
+          rec = TLastRecLoopBodyAbnormal() and
+          not c0 instanceof NormalCompletion and
+          not c0 instanceof ContinueCompletion and
+          not c0 instanceof BreakCompletion and
+          c = c0
+        )
+        or
+        // Last `catch` clause inherits throw completions from the `try` block,
+        // when the clause does not match
+        exists(SpecificCatchClause scc |
+          scc = cfe and
+          scc.isLast() and
+          throwMayBeUncaught(scc, c)
+        |
+          // Incompatible exception type: clause itself
+          result = scc
+          or
+          // Incompatible filter
+          result = lastSpecificCatchClauseFilterClause(scc)
+        )
+        or
+        cfe = any(TryStmt ts |
+            result = getBlockOrCatchFinallyPred(ts, c) and
+            (
+              // If there is no `finally` block, last elements are from the body, from
+              // the blocks of one of the `catch` clauses, or from the last `catch` clause
+              not ts.hasFinally()
+              or
+              // Exit completions ignore the `finally` block
+              c instanceof ExitCompletion
+            )
+            or
+            // If the `finally` block completes normally, it resumes any non-normal
+            // completion that was current before the `finally` block was entered
+            exists(NormalCompletion c0 |
+              result = lastTryStmtFinally(ts, c0) and
+              (
+                exists(getBlockOrCatchFinallyPred(ts, any(NormalCompletion nc))) and
+                c = c0
+                or
+                exists(getBlockOrCatchFinallyPred(ts, c)) and
+                not c instanceof NormalCompletion
+              )
+            )
+          )
       }
 
       pragma[nomagic]
-      private ControlFlowElement lastConditionalQualifiableExprChildExpr(
-        ConditionallyQualifiedExpr cqe, int i, Completion c
+      private ControlFlowElement lastRec(
```

Comment:
A comment on this predicate would be appreciated.

Reply:
We usually don't document folded predicates, but I can add a comment.